src/providers/VoiceFocusProvider/index.tsx

```diff
@@ -0,0 +1,170 @@
+  */
+  options?: VoiceFocusDeviceOptions;
+}
+
+interface VoiceFocusState {
+  isVoiceFocusSupported: boolean | undefined;
+  addVoiceFocus: (device: Device) => Promise<Device | VoiceFocusTransformDevice>;
+}
+
+const VoiceFocusContext = createContext<VoiceFocusState | null>(null);
+
+const VoiceFocusProvider: React.FC<Props> = ({
+  spec,
+  options,
+  children
+}) => {
+  const [isVoiceFocusSupported, setIsVoiceFocusSupported] = useState<boolean | undefined>(undefined);
+  const [voiceFocusDevice, setVoiceFocusDevice] = useState<VoiceFocusTransformDevice | null>(null);
+  const [voiceFocusTransformer, setVoiceFocusTransformer] = useState<VoiceFocusDeviceTransformer | null>(null);
+
+  // Check whether the value of spec is undefined or empty object
+  const vfSpec = useMemoCompare(spec, (prev: VoiceFocusSpec | undefined, next: VoiceFocusSpec | undefined): boolean => {
+    if (prev === next) {
+      return true;
+    }
+    if (next === undefined || Object.keys(next).length === 0) {
+      return true;
+    }
+    if (prev === undefined || Object.keys(prev).length === 0) {
+      return true;
+    }
+    return false;
+  });
+
+  const addVoiceFocus = useCallback(async (device: Device): Promise<Device | VoiceFocusTransformDevice> => {
+    if (!voiceFocusDevice) {
+      return createVoiceFocusDevice(device);
+    }
+    const vf = await voiceFocusDevice.chooseNewInnerDevice(device);
+    setVoiceFocusDevice(vf);
+    return vf;
+  }, [isVoiceFocusSupported]);
+
+  async function createVoiceFocusDevice(inner: Device): Promise<Device | VoiceFocusTransformDevice> {
+    if (!isVoiceFocusSupported) {
+      return inner;
+    }
+    try {
+      const transformer = await getVoiceFocusDeviceTransformer();
+      const device = await transformer?.createTransformDevice(inner);
+      if (device) {
+        setVoiceFocusDevice(device);
+        return device;
+      }
+    } catch (e) {
+      console.warn('Amazon Voice Focus is not supported.', e);
+    }
+    return inner;
+  }
+
+  async function getVoiceFocusDeviceTransformer(): Promise<VoiceFocusDeviceTransformer> {
+    if (voiceFocusTransformer) {
+      return voiceFocusTransformer;
+    }
+
+    const transformer = await VoiceFocusDeviceTransformer.create(spec, options);
+    setVoiceFocusTransformer(transformer);
+    return transformer;
+  }
+
+  let currentPromise: Promise<VoiceFocusDeviceTransformer>;
+
+  useEffect(() => {
+    async function initVoiceFocus() {
+      try {
+        const fetch = VoiceFocusDeviceTransformer.create(vfSpec, options);
+
+        /**
+        * We use currentPromise to store the latest promise of VoiceFocusDeviceTransformer.
+        * If the builder changes the spec or options when the previous promise is still pending,
+        * We will just grab the latest settings to create an Amazon Voice Focus transformer.
+        */
+        fetch.then((value) => {
```

Comment:
I think this should be part of `getVoiceFocusDeviceTransformer`, and we simply call it here and discard the resulting promise. Otherwise they're going to race.